www/waterfall_view/src/module/main.module.js

```diff
@@ -635,14 +640,89 @@ var WaterfallController = (function() {
             });
         }
 
-        renderNewData() {
+        toggleTag(tag) {
+            if (!this.$scope.tags_filter.includes(tag)) {
+                this.$scope.tags_filter.push(tag);
+            } else {
+                this.$scope.tags_filter = _.remove(this.$scope.tags_filter, function(currentTag) {
+                    return currentTag != tag;
+                });
+            }
+            this.$location.search("tags", this.$scope.tags_filter);
+            this.renderNewData(this.$scope.tags_filter);
+        }
+
+        isTagToggled(tag) {
+            if (this.$scope.tags_filter.includes(tag)) {
+                return 'label-success';
+            } else {
+                return 'label-default';
+            }
+        }
+
+        clearTags() {
+            this.$scope.tags_filter = this.tags_filter = [];
+            this.$location.search("tags", this.$scope.tags_filter);
+            this.renderNewData(this.$scope.tags_filter);
+        }
+
+        renderNewData(currentTags) {
+            currentTags = this.$location.search()['tags'];
+            if (currentTags != null) {
+                if (typeof currentTags == 'string'){
+                    this.$scope.tags_filter = this.tags_filter = [currentTags];
+                } else {
+                    this.$scope.tags_filter = this.tags_filter = currentTags;
```

Comment:
Are we guaranteed to get an array if `currentTags` is not a string?

Reply:
In fact I'm getting an object in the else case

Reply:
So tags_filter can be either array or object? sounds bad.

Reply:
`typeof tags_filter` is giving object but it's actually an array

Reply:
This is because javascript considers `typeof Array()` as `object`

Reply:
These issues would be solved if we use typescript